jax/experimental/jax2tf/tests/jax2tf_test.py

```diff
@@ -410,6 +415,238 @@ def outer(y):
+    self.CheckShapePolymorphism(add_all_jax, [(None, [("v, 4")]),
+                                              dict(a="(v, 4)", b=None)],
+                                input_signature=self.MakeInputSignature((["(4,)", "(4,)"], [("v, 4,")]),
+                                                                        dict(a="(v, 4)", b="(4,)")),
+                                expected_output_shape="(v, 4)")
+
+  def test_with_custom_vjp(self):
+    """Shape-polymorphic custom VJP."""
+    # TODO: is this test really adding anything???
+    @jax.custom_vjp
+    def f(x):
+      # x: [b1, b2, d1, d2] (a batch of matrices)
+      # res: [b1, b2, d1, d1]
+      return jnp.matmul(x, jnp.transpose(x, axes=(0, 1, 3, 2)))
+
+    # f_fwd: a -> (b, residual)
+    def f_fwd(x):
+      # x: [b1, b2, d1, d2]
+      # b: [b1, b2, d1, d1]
+      # residual: [b1, b2, d1, d2]
+      return f(x), 3. * x
+    # f_bwd: (residual, CT b) -> [CT a]
+    def f_bwd(residual, ct_b):
+      # residual: [b1, b2, d1, d2]
+      # ct_b: [b1, b2, d1, d1]
+      # ct_a: [b1, b2, d1, d2]
+      return jnp.matmul(ct_b, residual),
+
+    f.defvjp(f_fwd, f_bwd)
+    x = np.ones((2, 3, 4, 5), dtype=np.float32)
+    res_jax = f(x)
+    res_jax_grad = jax.grad(lambda x: jnp.sum(f(x)))(x)
+
+    f_tf = self.CheckShapePolymorphism(f, ["(batch1, batch2, d1, d2)"],
+                                       expected_output_shape="(batch1, batch2, d1, d1)")
+    self.assertAllClose(res_jax, f_tf(x))
+
+    xv = tf.Variable(x, dtype=dtypes.canonicalize_dtype(jnp.float_))
+    def tf_value_and_grad(xv):
+      with tf.GradientTape() as tape:
+        tape.watch(xv)
+        res_tf = f_tf(xv)
+        res_tf_grad = tape.gradient(res_tf, xv)
+        return res_tf, res_tf_grad
+
+    res_tf, res_tf_grad = tf_value_and_grad(xv)
+    self.assertAllClose(res_jax, res_tf)
+    self.assertAllClose(res_jax_grad, res_tf_grad)
+
+    # Now use TF tracing for the gradient
+    tf_grad = tf.function(
+      tf_value_and_grad,
+      autograph=False).get_concrete_function(
+        *self.MakeInputSignature("(batch1, batch2, 8, 9)"))
+    # The shape of the value
+    self.assertEqual((None, None, 8, 8), tuple(tf_grad.output_shapes[0]))
+    # The shape of the gradient should match the input
+    # TODO: there seems to be a bug here, the output should be (None, None, 8, 9)
+    # self.assertEqual((None, None, 8, None), tuple(tf_grad.output_shapes[1]))
+
+
+  def test_gradients_pytree(self):
+    """Shape polymorphism with gradients and pytrees for inputs and outputs."""
+    def f(x):
+      # x: dict(x=[b, 3, 4])
+      # res: dict(res=[b, 3, 4])
+      return dict(res=x["x"] * 2.)
+
+    f_tf = self.CheckShapePolymorphism(f, [dict(x=("b, 3, 4"))],
+                                       expected_output_shape=dict(res="(b, 3, 4)"))
+    x = dict(x=np.ones((2, 3, 4), dtype=np.float32))
+    xv = tf.Variable(x["x"], dtype=dtypes.canonicalize_dtype(jnp.float_))
```

Comment:
The dtype here should instead be `jnp.float32`. As it is, the test is not compatible with the x64 flag
```suggestion
    xv = tf.Variable(x["x"], dtype=jnp.float32)
```

Reply:
Fixed